streams/runners/storm/actions/src/main/java/com/hortonworks/streamline/streams/actions/storm/topology/TestTopologyDagCreatingVisitor.java

```diff
@@ -0,0 +1,147 @@
+        this.sourceToReplacedTestSourceMap = new HashMap<>();
+        this.sinkToReplacedTestSinkMap = new HashMap<>();
+    }
+
+    @Override
+    public void visit(RulesProcessor rulesProcessor) {
+        visit((StreamlineProcessor) rulesProcessor);
+    }
+
+    @Override
+    public void visit(StreamlineSource source) {
+        String id = source.getId();
+        String sourceName = source.getName();
+
+        if (!testRunSourcesForEachSource.containsKey(sourceName)) {
+            throw new IllegalStateException("Not all sources have corresponding TestRunSource instance. source name: " + sourceName);
+        }
+
+        Config config = new Config(source.getConfig());
+
+        TestRunSource testRunSource = testRunSourcesForEachSource.get(sourceName);
+
+        testRunSource.setId(id);
+        testRunSource.setName(sourceName);
+        testRunSource.setConfig(config);
+        testRunSource.setTransformationClass(TestRunSourceSpoutFluxComponent.class.getName());
+
+        testTopologyDag.add(testRunSource);
+        sourceToReplacedTestSourceMap.put(sourceName, testRunSource);
+    }
+
+    @Override
+    public void visit(StreamlineSink sink) {
+        String id = sink.getId();
+        String sinkName = sink.getName();
+
+        if (!testRunSinksForEachSink.containsKey(sinkName)) {
+            throw new IllegalStateException("Not all sinks have corresponding TestRunSink instance. sink name: " + sinkName);
+        }
+
+        Config config = new Config(sink.getConfig());
+
+        TestRunSink testRunSink = testRunSinksForEachSink.get(sinkName);
+
+        testRunSink.setId(id);
+        testRunSink.setName(sinkName);
+        testRunSink.setConfig(config);
+        testRunSink.setTransformationClass(TestRunSinkBoltFluxComponent.class.getName());
+
+        testTopologyDag.add(testRunSink);
+        sinkToReplacedTestSinkMap.put(sinkName, testRunSink);
+
+        copyEdges(sink);
+    }
+
+    @Override
+    public void visit(StreamlineProcessor processor) {
+        testTopologyDag.add(processor);
+
+        copyEdges(processor);
+    }
+
+    private void copyEdges(InputComponent inputComponent) {
+        List<Edge> edgesTo = originTopologyDag.getEdgesTo(inputComponent);
+        edgesTo.forEach(e -> {
+            OutputComponent from = e.getFrom();
+            InputComponent to = e.getTo();
+            Edge newEdge = new Edge(e.getId(), e.getFrom(), e.getTo(), e.getStreamGroupings());
+
+            StreamlineSource replacedSource = sourceToReplacedTestSourceMap.get(from.getName());
+            StreamlineSink replacedSink = sinkToReplacedTestSinkMap.get(to.getName());
+
+            if (replacedSource != null) {
+                newEdge.setFrom(replacedSource);
+            }
+
+            if (replacedSink != null) {
+                newEdge.setTo(replacedSink);
+            }
+
+            testTopologyDag.addEdge(newEdge);
+        });
+    }
+
+    @Override
+    public void visit(Edge edge) {
+        // no-op
```

Comment:
I gave a first try with that, but realized it's not easy.
To copy Edge we need to know about source and destination, and if either (or both) source or destination is source or sink, it should connect with replaced thing, which is not easy as it seems.